teach fsharp to dotnet new

enricosada opened the pull request:
ref #647 

use argument `-f#` or `--fsharp` ( and `-c#` or `--csharp` ).
Default without arguments is csharp, like before.

the generated f# program doesnt restore correctly, that's because of Microsoft/visualfsharp#832 but it's not a problem atm, it need only a bump of version after it's fixed.

dnfclas commented:
Hi __@enricosada__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

krwq commented on src/Microsoft.DotNet.Tools.New/Program.cs:
```diff
@@ -79,8 +99,21 @@ public static int Main(string[] args)
             app.Description = "Initializes empty project for .NET Platform";
             app.HelpOption("-h|--help");
 
+            var csharp = app.Option("-c#|--csharp", "Create an C# project", CommandOptionType.NoValue);
```
I think that explicit form should be either (`dotnet new --lang csharp`, `dotnet new --lang fsharp`) or (`dotnet new c#`, `dotnet new csharp`, ...)

I'm also wondering to maybe provide the source file name itself and if someone provide extension .cs then use csharp template and if .fs then use fsharp template. It would be something like `dotnet new Program.cs`. cc @piotrpMSFT @davidfowl @blackdwarf @anurse @brthor @Sridhar-MS

brthor replied:
I agree here, adding a new option for every language is going to be ugly eventually IMO.

I like `--lang` as there is a default (csharp) and it makes it clear this is an optional value.

enricosada replied:
Ok, i'll add `--lang`
Nice idea the Program.fs

krwq replied:
@brthor, we could technically support both options. `dotnet new --lang fsharp` sounds pretty long to type so I'd still consider: `dotnet new fsharp` considering that this is first user experience. Also: should we default to csharp? If `dotnet new` displayed small usage help so you could copy&paste the command it wouldn't be too bad but:
`dotnet new --help` + `dotnet new --lang fsharp` does not sound too fun

brthor replied:
`dotnet new fsharp/csharp` seems fine to me. Given most users of the toolchain will be using csharp, it seems like a reasonable default. In 80% of cases, the only thing you'd need to type is `dotnet new`

krwq commented on src/Microsoft.DotNet.Tools.New/Program.cs:
```diff
@@ -24,13 +30,27 @@ private static string GetFileNameFromResourceName(string s)
         }
 
-        public int CreateEmptyProject()
+        public int CreateEmptyProject(Language language)
```
I'm not 100% sure if I like the idea the idea of:
- having string input
- converting that to enum
- converting that to different string
- generating template from that different string.

Can we just use the original string instead and map it to directory name?

enricosada replied:
Ok, i'll use `C#` and `F# directories and remove enum/conversion

brthor replied:
The enum conversion does seem a little weird, but directly using input as a directory name seems dangerous to me. I can't think of a situation where this causes a vulnerability, but directory traversal via user input has been known to cause scary things :smile:

enricosada replied:
I prefer to make types than string interpolation

There embedded files are used, no problems about directory traversal ( all parts are dot separated as a single name )

I'll whitelist strings ( c# and f# ) for safety

It's ok use string directly, we are using these only to get the directory name of resources, so less translation string -> enum -> string

brthor replied:
After more thought, we have many places where we expect a full path as input, so I wouldn't worry about that.

krwq replied:
Well, attacker won't get anything they can't already get so I guess it's ok :) Whitelist sounds good to me (some mappings csharp->c# or reverse sound good too)

enricosada commented:
so:

- added `--lang F#`. Default to `C#` ( so `dotnet new` = `dotnet new --lang csharp` )
- `C#`, `csharp` and `cs` aliases works too. Same for `F#`, `fsharp`, `fs` ( case insensitive )
- removed enum, list aliases and error if argument not supported
- fix doc

piotrpMSFT commented:
@blackdwarf @davidfowl can you take a look?

piotrpMSFT commented on src/Microsoft.DotNet.Tools.New/Program.cs:
```diff
@@ -79,8 +77,27 @@ public static int Main(string[] args)
             app.Description = "Initializes empty project for .NET Platform";
             app.HelpOption("-h|--help");
 
+            var lang = app.Option("-l|--lang <LANGUAGE>", "Language of project ( C#, F# )", CommandOptionType.SingleValue);
```
C# | F#?

blackdwarf replied:
True "|" is the separator character that needs to be used here.

enricosada replied:
sure, `[c#|f#]` it's ok?

blackdwarf replied:
Yeah. I'm thinking, should we remove the "c#" option and just leave "csharp" and "cs" (same for F#)? makes more snese...

piotrpMSFT commented on src/Microsoft.DotNet.Tools.New/README.md:
```diff
@@ -12,15 +12,26 @@ The new command provides a convenient way to initalize a valid .NET Core project
+`dotnet new --lang f#`
+>Drops a sample F# project in the current directory.
+
+`dotnet new --lang csharp`
```
C#, righ?

enricosada replied:
It's the same, `csharp` or `c#` or `cs` are alias. i wrote `csharp` to make it explict

piotrpMSFT commented:
I'd like to see this concept generalized. The --lang option suggests that all 'new' outputs will be pivotable on language. If we end up adding a "New ASP.Net" starter project then are we going to provide that in C# and F#? What happens when VB comes along?

My preference would be to enable named templates, wherein the CSharpConsole template is the default. This gives us the flexibility to expand the generatable templates along a single axis [template name] instead of dealing with multidimensional choices [along with the error conditions those would need to handle].

The code should also be data-driven from the templates. For example, we could assume that the root directory of each template is the template name. If a concept of aliases is desirable then we can think about how to encode that. This command is picking a template and writing it to disk... let's try to reduce complexity.

Adding a third template should be as easy as adding more to the resources of this command.

blackdwarf commented:
@piotrpMSFT when we add VB, we add it for all the types that exist and where it makes sense. 

@enricosada as per #655 can you also add --type option as was defined there? Then this would make more sense and would go towards the plan we have for the command all-up. The support for creating the directory is still not specc-baked. :)

piotrpMSFT commented:
I hadn't seen #655 before commenting on this issue. I see we have a broader plan than what I saw before commenting :)

enricosada commented:
@blackdwarf can i do `--type` as a separate pr? in this one i want to add `--lang` only.
i really like do small pr, less problems about merge issue, less scope, faster feedback loop.

btw i think in the `--type` pr we can rename the directory from `CSharpTemplate` to `CSharpConsole`, so we can concat language name + template name

enricosada commented:
> The --lang option suggests that all 'new' outputs will be pivotable on language. If we end up adding a "New ASP.Net" starter project then are we going to provide that in C# and F#? What happens when VB comes along?

If not supported, we can write an error

```
Template web not supported for F#.

Supported templates for F# are:

- console
- library
- suave
```

etc, reading the list from resources templates, not too bad.

yes, i think is a good idea `LanguageName` ( --lang )  + `TemplateName` ( --type ) => directory inside resources.
so `c#` + `web` => `csharp-web`

blackdwarf commented:
@enricosada you can, of course, but would like to merge them at similar timeframes, since I would like to evolve the tool forward in an streamlined fashion. 

Also, I think @piotrpMSFT comment also had to do with how the tool would surface this to the users. As in, we would not have the --type and --lang switches, but rather have something like `dotnet new csharp_console` and it would simply yank out the corresponding template and drop it in a directory.

enricosada commented:
@blackdwarf i'll add `--type` to this one so, i need to fix merge conflict anyway :smile: 

i think `dotnet new csharp_console` it's good, but it's not easier `dotnet new csharp console`? why use `_`? it's easier to implement, but it's bad for users

`dotnet new [language] [type]` ( `dotnet new [language] [type] [directoryname]` works too )
But like that write `csharp` is pretty much required, because defaults sucks

Maybe like lein is ok
`dotnet new [type] [directory] [--lang]` both `type` and `--lang` optional, so `dotnet new console` default to c#

blackdwarf commented:
@enricosada good points. I like commands to read out as sentences because I think people than have an easier time learning them and getting into the habit of using them. 

Since we are also planning to add directory name, we need to take that into account when thinking about the syntax of the command. So `dotnet new [type] [lang] [dirname]` seems like an overkill. Switching the params place would work, i.e. `dotnet new [dirname] [type] [lang]`. But params seem like a better option to me. Of course, I could be wrong. :)

enricosada commented:
`dotnet new hello_world console` it's ok, `dotnet new hello_world` ok too

`dotnet new hello_world console fsharp` it's not too bad

blackdwarf commented:
True, its not. Again, am more partial to parameters on invocation given that they specify meaning much better. Here, if I switch the arguments (as an user) I can get errors or weird results.

piotrpMSFT commented:
with params, am I correct in understanding that:

default --lang = "csharp"
default --type = "console"

?

enricosada commented:
For --lang yes, default is csharp now
When i add --type default will be console

blackdwarf commented:
@piotrpMSFT yes.

piotrpMSFT commented:
ok, sounds like once @enricosada rebases we can merge.

enricosada commented:
@piotrpMSFT i'll do asap, now i can build :smile:

piotrpMSFT commented:
@enricosada if this is intended to go into 1.0, please retarget the PR to rel/1.0.0.

enricosada commented:
@piotrpMSFT sry i didnt see anything about branch targets. I'd like to have it in next drop (`rc2` ?), so `master` or `rel/1.0.0`? `master` is after `1.0`?
The rebase is obv not a problem, just asking because i dont know where you want this feature, or what branch mean

enricosada commented:
ok added `--type`

- `dotnet new` => without arguments, same behaviour, a c# console
- `dotnet new --lang f#` => default f# type => f# console
- `dotnet new --lang f# --type console` => f# console
- `dotnet new --type console` => default lang c# => c# console
- `dotnet new --type invalid` => error with list of types (default c#)
- `dotnet new --type invalid --lang f#` => error with list of types for f#

enricosada commented:
@piotrpMSFT sry i didnt see anything about branch targets. I'd like to have it in next drop (i see tag `RC2-Feb` ?), so `master` or `rel/1.0.0`? `master` is after 1.0?
The rebase is obv not a problem, just asking because i dont know where you ms team want this feature, or what branches mean

enricosada commented:
@piotrpMSFT i'll rebase to branch `rel/1.0.0`

enricosada commented:
created another pr #1027 (same code) because cant change target branch in github
